src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+            File speciesUniprotFile = createNewFile(this.destination + biomartSpeciesName + "_uniprot");
+            for (String uniprotIdLine : uniprotIdLines) {
+                Files.write(Paths.get(speciesUniprotFile.toString()), uniprotIdLine.getBytes(), StandardOpenOption.APPEND);
+            }
+        }
+    }
+
+    private String getMicroarrayTypesQuery(String biomartSpeciesName) {
+        return this.getDataURL() + microarrayTypesBaseQuery.replace("BIOMART_SPECIES_NAME", biomartSpeciesName);
+    }
+
+    /**
+     * Query Biomart for Microarray 'identifiers' associated with the microarray type
+     * @param biomartSpeciesName -- String of species name in Biomart format (eg: hsapiens)
+     * @param microarrayTypes -- String, type of microarray identifiers being queried for.
+     * @throws Exception
+     */
+    private void retrieveAndStoreMicroarrayData(String biomartSpeciesName, String biomartXMLPath, Set<String> microarrayTypes) {
+
+        // Query Biomart for microarray identifiers associated with microarray type.
+        for (String microarrayType : microarrayTypes) {
+            try {
+                logger.info("Retrieving data associated with microarray type " + microarrayType);
+                Set<String> microarrayResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartXMLPath, biomartSpeciesName, microarrayType), 0);
+                Set<String> microarrayIdentifierLines = new HashSet<>();
+                for (String microarrayResponseLine : microarrayResponseLines) {
+                    List<String> tabSplit = Arrays.asList(microarrayResponseLine.split("\t"));
+                    // Microarray ID is 4th column, and sometimes it isn't in a line.
+                    // Since we need the ID, we require the List to have 4 elements.
+                    if (tabSplit.size() == 4) {
+                        microarrayIdentifierLines.add(microarrayResponseLine + "\n");
+                    }
+                }
+
+                // Write data to file.
+                if (!microarrayIdentifierLines.isEmpty()) {
+                    File speciesMicroarrayFile = createNewFile(this.destination + biomartSpeciesName + "_microarray");
+                    for (String probeIdLine : microarrayIdentifierLines) {
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
```

Comment:
Reminder (we just spoke in person, this is to write it down): this code _will_ fail for species (frog, pig) that are known to not have data for certain types of queries. Add a comment in the code where the exception gets thrown (or even better, add it to the Exception's message) so that people know that for _some_ species, an exception is probably normal (until Ensembl adds data for those species - if they ever do).

Reply:
Yea I'll do both actually, add a comment in the code that is more thorough, but the Exception message will also say something along the lines of 'This is typical for Frog, D. discoideum, etc